### DISCUSSION TITLE
implementing Rank 1 constraint

### DISCUSSION THREAD

**Discussion by Prashantgrid** [cvxpy/cvxpy#1343] [Q&A] [answered]
Dear cvxpy team,

I am trying to model optimal power flow with SDP relaxation, to retrieve  solutions the PSD matrix should be rank 1. since I am dropping the rank constraint I am getting a full rank matrix, which is totally against the assumptoin of rank -1 solution. 

Is there any way to implement rank constraints? 

Regards,
Prashant

**Comment by Prashantgrid** [comment]
I already tried rank and trace equivalence constraint, but it makes the problem infeasible. Maybe someone have a better idea?

Thanks

**Comment by rileyjmurray** [comment, marked as answer]
The set of rank-1 matrices is not convex even under very restrictive assumptions. It's easy to see why: given any two linearly independent vectors ``v1, v2``, the matrices ``mat1 = np.outer(v1,v1)`` and ``mat2 = np.outer(v2, v2)`` are rank-1, but ``mat3 = weight*mat1 + (1-weight)*mat2`` is rank-2 for every "weight" not equal to zero or one. This nonconvexity is precisely why you need to approach AC OPF with an SDP relaxation as opposed to an exact SDP reformulation.

For many OPF problems seen in practice, the optimal solution to the traditional SDP relaxation just so happens to have low-rank. Note that the solution may have rank > 1 and the SDP relaxation can still be exact -- this happens when the OPF problem has isolated globally optimal solutions. In order to recover optimal solutions from a rank > 1 matrix, you can look at Jean-Bernard Lasserre's book "An Introduction to Polynomial and Semi-Algebraic Optimization." 

Unfortunately, it is a pain to implement these things from scratch. The polynomial optimization community hasn't prioritized making their techniques accessible by Python. [I personally have a Python package for convex relaxations to polynomial optimization problems](https://rileyjmurray.github.io/sageopt/), but it isn't well-suited to low-degree problems like AC OPF.

**Reply by Prashantgrid** [reply]
Thanks for the reply. I found the problem, when I was checking the eigen values, some of them were too samll in order of e-5 but they were considered as non-zero and I got full rank everytime. Thanks for the valuable information :)
